rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,710 @@
+
+   - When breaking changes are introduced, the major version number is incremented.
+
+     Examples include semantic changes or removal of `builtins`, operators, or syntactic constructs.
+
+   <details><summary>Arguments</summary>
+
+   - (+) The principled solution: guarantees reproducibility given a fixed language version
+   - (+) Makes explicit current practice of adding features to the language, and allows introducing breaking changes in a controlled fashion, which is currently not possible at all
+   - (-) Implies additional overhead in development effort:
+     - Either for Nix maintainers to accommodate that practice in the release lifecycle
+       - For example, one would have to batch language changes for a major version bump to limit the number of increments
+         - (+) This would be beneficial for alternative implementations in terms of churn and effort to keep up
+     - Or implementors of alternative evaluators catching up with changes
+       - (+) Specifying the language precisely via the version actually offers alternative implementations an alternative to catching up: only support a given language version
+   - (+) This essentially nudges one to organise Nix language (specification or evaluator implementation) development to be more independent of the rest of Nix
+     This is good, since it in turn forces stronger separation of concerns and more architectural clarity
+   - (-) Prohibits best-effort attempts at evaluating expressions with possibly incompatible evaluators
+     - (+) With the proposed level of strictness, one doesn't have to rely on best effort but can instead be explicit
+   - (-) Fixing evaluator bugs (i.e., clearly unintended behavior) after releases would technically require a version bump and therefore (theoretically) cooperation by expression authors
+     - This could be communicated with an increment in the Nix patch-level version, as is already practice
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Bump major version only when evaluation result on prior version evaluators would be *substantially* different
+     - (+) Leaves room for judgement by developers
+       - (+) Allows controlling progression of versions to some degree
+     - (-) Conversely, leaves room for sneaking in breaking changes unannounced
+       - (-) This loses compatibility guarantees we'd get from a stricter paradigm
+       - (-) Deprives expression authors of ability to be selective with evaluator versions
+     - (-) Due to hashing this is often not much different from taking *any* change into account
+
+   </details>
+
+1. Semantics are preserved across file boundaries for past language versions.
+
+   This should be fairly straightforward to implement since values passed around in the evaluator carry all the information needed to force them.
+   Newer parts of the evaluator can always wrap their values in interfaces that are accepted by older parts, as far as possible.
```

Comment:
Yikes, this sounds extremely heavyweight!  You mean we're going to wrap individual thunks and integers with an additional language-version parameter?

I think this is not a great idea.

Also the example isn't terribly persuasive.  If Nix were to add complex numbers, surely it would also extend the definition of addition to cover complex-float addition.  As long as old-version code isn't able to *create* complex numbers, the behavior of a 100%-old-code codebase can't change.  So there's no harm in evaluating old code with the newer version of `+` that understands complex numbers.

Reply:
The idea would be to do so at parsing time, so that no additional tracking should be needed in the evaluator. Also this would only be done to the extent necessary, integers for example are unlikely to ever get hit.

I think it is fair to have some performance overhead with cross-version code, as long as code that does not make use from it does not pay a performance penalty. Something something zero-cost …

Reply:
> The idea would be to do so at parsing time

That... isn't what the RFC proposes.

Have you tried actually implementing this part of the proposal?  I really think you should, before it is approved.  You are likely to be quite surprised.

Reply:
I'm not the RFC author, but at least this is what I had in mind which resulted in the text above. And I agree that some significant implementation effort will have to be conducted before approving this RFC.